src/java/com/palantir/cassandra/cvim/CrossVpcIpMappingHandshaker.java

```diff
@@ -48,14 +54,136 @@
     private static final DebuggableScheduledThreadPoolExecutor executor = new DebuggableScheduledThreadPoolExecutor(
     "CrossVpcIpMappingTasks");
     private volatile ScheduledFuture<?> scheduledCVIMTask;
-    public final static Duration interval = Duration.ofSeconds(1);
+    public final static Duration scheduledInterval = Duration.ofSeconds(1);
+    public final static Duration minHandshakeInterval = Duration.ofMillis(500);
+    private static volatile long lastTriggeredHandshakeMillis = 0;
+    private final ConcurrentHashMap<InetAddressIp, InetAddressIp> privatePublicIpMappings;
+    private final ConcurrentHashMap<InetAddressIp, InetAddressHostname> ipHostnameMappings;
 
-    private CrossVpcIpMappingHandshaker() {}
+    private CrossVpcIpMappingHandshaker()
+    {
+        this.privatePublicIpMappings = new ConcurrentHashMap<>();
+        this.ipHostnameMappings = new ConcurrentHashMap<>();
+    }
 
-    public void triggerHandshakeWithSeeds()
+    public void updateCrossVpcMappings(InetAddressHostname host, InetAddressIp internalIp, InetAddressIp externalIp)
     {
+        if (!DatabaseDescriptor.isCrossVpcInternodeCommunicationEnabled()) {
+            return;
+        }
+        InetAddressIp oldExternalIp = this.privatePublicIpMappings.get(internalIp);
+        if (!externalIp.equals(oldExternalIp))
+        {
+            this.privatePublicIpMappings.put(internalIp, externalIp);
+            logger.trace("Updated private/public IP mapping for {} from {}->{} to {}",
+                         host, internalIp, oldExternalIp, externalIp);
+        }
+
+        InetAddressHostname old = this.ipHostnameMappings.get(internalIp);
+        if (!host.equals(old))
+        {
+            this.ipHostnameMappings.put(internalIp, host);
+            logger.trace("Updated private IP/hostname mapping from {}->{} to {}", internalIp, old, host);
+        }
+        old = this.ipHostnameMappings.get(externalIp);
+        if (!host.equals(old))
+        {
+            this.ipHostnameMappings.put(externalIp, host);
+            logger.trace("Updated public IP/hostname mapping from {}->{} to {}", externalIp, old, host);
+        }
+    }
+
+    /**
+     * Depending on which cross-vpc flags are enabled/disabled, will check the cross-vpc mappings and either swap the
+     * given endpoint with an endpoint derived from a public IP, or DNS using a hostname. If no mappings are found will
+     * return the original endpoint.
+     */
+    public InetAddress maybeSwapAddress(InetAddress endpoint)
+    {
+        if (!DatabaseDescriptor.isCrossVpcInternodeCommunicationEnabled()) {
```

Comment:
can we add `!DatabaseDescriptor.isCrossVpcHostnameSwappingEnabled()` up here and remove it below? seems like if its false then we always return "endpoint"

Reply:
oh nevermind, i'm reading this wrong.